kimchi/src/circuits/polynomials/foreign_mul/witness.rs

```diff
@@ -0,0 +1,211 @@
+//! Witness computation
+
+/// TODO: make dynamic
+use ark_ff::PrimeField;
+use array_init::array_init;
+use num_bigint::BigUint;
+use o1_utils::FieldHelpers;
+
+use crate::circuits::polynomial::COLUMNS;
+
+// The maximum supported foreign field element size is 264-bits
+const MAX_LIMBS: usize = 3;
+const LIMB_SIZE: usize = 88;
+
+enum WitnessCell {
+    Copy(CopyWitnessCell),
+    Limb,
+    Sublimb(SublimbWitnessCell),
+    Zero,
+}
+
+// Witness cell copied from another
+struct CopyWitnessCell {
+    row: usize,
+    col: usize,
+}
+impl CopyWitnessCell {
+    const fn create(row: usize, col: usize) -> WitnessCell {
+        WitnessCell::Copy(CopyWitnessCell { row, col })
+    }
+}
+
+// Witness cell for a foreign field element limb
+struct LimbWitnessCell;
+impl LimbWitnessCell {
+    const fn create() -> WitnessCell {
+        WitnessCell::Limb
+    }
+}
+
+// Witness cell for a foreign field element sub-limb
+struct SublimbWitnessCell {
+    row: usize,   // Cell row
+    col: usize,   // Cell col
+    start: usize, // Starting bit offset
+    end: usize,   // Ending bit offset (exclusive)
+}
+impl SublimbWitnessCell {
+    // Params: source (row, col), starting bit offset and ending bit offset (exclusive)
+    const fn create(row: usize, col: usize, start: usize, end: usize) -> WitnessCell {
+        WitnessCell::Sublimb(SublimbWitnessCell {
+            row,
+            col,
+            start,
+            end,
+        })
+    }
+}
+
+// An cell containing zero
+struct ZeroWitnessCell;
+impl ZeroWitnessCell {
+    const fn create() -> WitnessCell {
+        WitnessCell::Zero
+    }
+}
+
+// Generate witness in shape that constraints expect (TODO: static for now, make dynamic)
+const WITNESS_SHAPE: [[WitnessCell; COLUMNS]; 4] = [
```

Comment:
This spells out the layout of the gates much more clearly than the current code for the gates does. Is it possible to re-use / adapt this representation for that?

Reply:
I'm not sure.  I guess this would be an alternative format to using the expr framework or some kind of wrapper around it.

It's an interesting idea.  Perhaps I could traverse this structure in the `constraints()` function and auto-generate the constraints from it.  It's a cool idea, but would be a big change.  I propose to do it in the subsequent PR.

Also, it feels like it would be good to re-use the same code for both witness generation and generating the constraints.  I think this is very similar to the code you and Izzy have work on for creating circuits, which I would rather use than re-invent.  So maybe it makes more sense I convert things to the new circuit interface once it's ready?

Reply:
Note that Izzy's circuit DSL is for creating circuits, not laying out individual gates. I don't think you'll be able to use it for this.

Reply:
Interesting.  So something similar to that for specifying the gate layout and generating constraints perhaps.

Reply:
This will be done in a subsequent PR.